crates/custody/src/threshold.rs

```diff
@@ -108,6 +109,65 @@ pub async fn follow(config: &Config, terminal: &impl Terminal) -> Result<()> {
     Ok(())
 }
 
+/// A distributed key generation protocol, producing a config without a centralized dealer.
+///
+/// Unlike the deal method on Config, this method will never have any participant know
+/// the key. Otherwise, the parameters controlling the threshold and the number of participants
+/// are the same as that method.
+///
+/// This takes in a terminal, because it requires interacting with the other participants.
+pub async fn dkg(t: u16, n: u16, terminal: &impl Terminal) -> Result<Config> {
+    let expected_responses = n.saturating_sub(1) as usize;
+    // Round 1 top
+    let (round1_message, state) = dkg::round1(&mut OsRng, t, n)?;
+    terminal
+        .explain("Round 1/2: Send this message to all other participants:")
+        .await?;
+    terminal.broadcast(&to_json(&round1_message)?).await?;
+    // Round 1 bottom
+    terminal
+        .explain(&format!(
+            "Round 1/2: Gather {expected_responses} messages from the other participants:"
+        ))
+        .await?;
+    let round1_replies = {
+        let mut acc: Vec<dkg::Round1> = Vec::new();
+        while acc.len() < expected_responses {
+            let string = terminal
+                .next_response()
+                .await?
+                .ok_or(anyhow!("expected message from another participant"))?;
+            acc.push(from_json(&string)?);
+        }
+        acc
+    };
+
+    // Round 2 top
+    let (round2_message, state) = dkg::round2(&mut OsRng, state, round1_replies)?;
+    terminal
+        .explain("Round 2/2: Send this message to all other participants:")
+        .await?;
+    terminal.broadcast(&to_json(&round2_message)?).await?;
+    // Round 2 bottom
+    terminal
+        .explain(&format!(
+            "Round 2/2: Gather {expected_responses} messages from the other participants:"
+        ))
+        .await?;
+    let round2_replies = {
+        let mut acc: Vec<dkg::Round2> = Vec::new();
+        while acc.len() < expected_responses {
```

Comment:
Here we're checking that we're getting the right number of responses, should we also be checking at each phase that we have no duplicate vks and that our own vk is not provided in a response?

Reply:
Yeah, we should do this internally, and also in signing